### PR TITLE
Look for symmetrical position in cache.

### PR DESCRIPTION
**Pull request by TFiFiE**
As suggested in https://github.com/gcp/leela-zero/issues/1084#issuecomment-375668610.

I took the liberty of changing the benchmark position to one in which this doesn't cause a speedup, because it's not typical for the game as a whole.

### CODE REVIEW

**Comment by gcp** [comment]
How does the performance look?

**Comment by TFiFiE** [comment]
This gives a speed boost for the first three moves or so and any performance hit afterwards is too insignificant for me to detect.

**Comment by gcp** [comment]
The other PR broken this, but I'll need a bit to review this anyway, wrt calc_hash changes, performance impact and the need for the rotation table (which relates to performance impact again).

**Comment by gcp** [comment]
@TFiFiE can you do a me a favor and rebase this?

File: src/FastBoard.h [hunk 1]
```diff
@@ -125,6 +125,8 @@ class FastBoard {
     int m_boardsize;
     int m_squaresize;
 
+    std::array<std::array<unsigned short, MAXSQ>, 8> m_symmetry_idx;
```
Comment (Hersmunch) [inline src/FastBoard.h]: How much of a pain would it be to make it so there is only one of these tables? Or is there a reason why it makes sense to have one for each object?

Comment (TFiFiE) [reply src/FastBoard.h]: It's tied to the board size, which is variable for this class.

Comment (Hersmunch) [reply src/FastBoard.h]: Of course. Can totally see it now, thanks.

Comment (gcp) [reply src/FastBoard.h]: If we can live with an extra indirection you can make it a static class owned vector (indexed by board size) and just check if that is initialized the moment you initialize FastBoard.

Alternatively, do the calculation on the fly...if necessary limit the symmetry check to the first 10-20 moves or so?

That's some of the things I want to look at.

File: src/Network.cpp [hunk 2]
```diff
@@ -868,8 +870,19 @@ Network::Netresult Network::get_scored_moves(
 
     if (!skip_cache) {
         // See if we already have this in the cache.
-        if (NNCache::get_NNCache().lookup(state->board.get_hash(), result)) {
-            return result;
+        for (auto symmetry = 0; symmetry < 8; ++symmetry) {
```
Comment (Hersmunch) [inline src/Network.cpp]: Name clash now that `rotation` parameter has been renamed. Sorry about that..

**Comment by AncalagonX** [comment]
I have been running this in my custom LZ build (essentially mirrors LZ_next + LCB_max_root) for a week now with Lizzie over many, many 9dan vs. LZ games.

I have noticed no downsides or performance decreases. In the beginning, it massively boosts reported visit counts in all four corners as expected.

Although I cannot personally confirm one way or the other whether the implementation has any bugs in aggregating symmetrical moves or truly results in better move selection, I can say that: 1) I have noticed no errors, misplays, or bizarrities from enabling this PR, which presumably is in effect for the entire game. 2) I have noticed no decrease in n/sec overall, although for this point I have only conducted the bare minimum of testing.

If there are no identifiable problems in how this is implemented, this PR could be enabled for selfplay and provide a mild boost to early-game search strength.

Though **do note that** -v 3200 might interfere with the greatly boosted visit counts from this PR—and so perhaps a special allowance of extra visits for early game symmetrical positions should be given, or else a "true visits" value should be tallied separately from the reported "true+symmetrical position visits" search results for the purposes of correctly enforcing a full search with 3200 real visits. Otherwise, using the visit counts directly from this PR might result in early-game searches erroneously ending search calculations hundreds or thousands of visits earlier than the 3200 mark.

This is a simple issue to address in any case, but I cannot fully interpret the code above on my own so perhaps this effect has already been accounted for.

**Comment by gcp** [comment]
This doesn't interfere with visit counting AFAIK, just makes it faster.

I basically only want to see if we can get the extra array out of FastBoard, and if getting rid of it would ( to keep performance parity) require limiting this to the first moves of the game.

**Comment by TFiFiE** [comment]
Why would you want such a change if only one `FastBoard` object is typically constructed and `FastBoard::reset_board` isn't called all the time?

**Comment by AncalagonX** [comment]
I've only conducted rather shallow testing on the issue, but I see no performance impacts in long games with this enabled.

That said, it's pretty clear that there is negligible benefit beyond the first x number of moves. I'd be happy with a max of 10 moves, just judging by gut feelings. If we are limiting this, considerations may need to be made for symmetrical play involving pre-placed handicap stones depending on how LZ counts which move it's on internally.

**Comment by gcp** [comment]
>if only one FastBoard object is typically constructed

Because it might be quite useful to construct multiple, for example if asked to list dead stones.

**Comment by gcp** [comment]
I guess this should be disabled during training? Else we can't take advantage of rotational asymmetry in the opening.

**Comment by gcp** [comment]
I reworked this a bit in #1275.

**Comment by gcp** [comment]
I can't measure a performance difference, and the other pull addresses the self-play games, so going to close this.

**Comment by TFiFiE** [comment]
@gcp: Can you reopen this?

**Pull request closed without merging**